### ISSUE TITLE
Creating backup still not working with v4.20

### ISSUE THREAD

**Issue by PiCob** [signalapp/Signal-Android#7831]
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
Well, creating backups is still not working for me. If I try to, it counts up the number of messages (> 20.000) and shows a motivation bar icon that a backup is being created, but a little later the notification disappears and a look in the backup options says: "Last backup: Never" (apparently due to a Java.io.IOException, see debug log)

`java.io.IOException: Renaming temporary backup file failed!`

Related: #7692 (but that one has been closed) - feel free to close this one if #7692 gets reopened.

FWIW: I do have an external SD card in my phone.

### Steps to reproduce
- Create backup

**Actual result:** Starts creating a backup and fails after some time
**Expected result:** Backup is created and saved successfully.

### Screenshots
Not needed.

### Device info

**Device:** Moto G4 Plus
**Android version:** 7.0.0 (stock)
**Signal version:** 4.20.4

### Link to debug log
https://debuglogs.org/35ddef38e5940589a4a7787013e0aa2a78ab33194f1df156878efa2af54c9637

**Comment by nextgen-networks** [comment]
Backup is not working for me since Saturday (assume it was the 4.20.2 update).

**Device info**
Device: Xiaomi Mi MIX 2
Android version: 8.0.0 (xiaomi.eu 8.5.8 rom)
Signal version: 4.20.4

**Link to debug log** https://debuglogs.org/990f25c1e137b590b2336114475b698454412a704ec2d2052dd50494dffd8c01

**Comment by Riamse** [comment]
Same problem as described above, on Signal 4.19.3.

**Comment by Riamse** [comment]
I literally don't think in my 3+ years of using Signal that the backup functionality has EVER worked. Is there something about the concept of encrypted backups from encrypted conversations that makes it a particularly hard problem?

**Comment by 0rC0** [comment]
Same problem for me:

Device info
Device: Motorola XT1524 (Moto E LTE (2nd gen))
Android : 7.1.2 (5fb4f458f4, lineage_surnia-userdebug 7.1.2 NJH47F 5fb4f458f4)
Signal version: 4.19.3

**Comment by minnmann** [comment]
@Riamse A full encrypted backup has only been implemented in Signal 4.17 (March 2018), mostly with commit 24e573e537639f6f8ff40fd774cf9ff079bbacce. 

I would guess, there are two different problems:

@PiCob The problem in your case seems to be the external storage. Did you try to eject the SD card and then run the backup? This workaround at least worked in 4.19.

@nextgen-networks seems to have a different problem as there is another exception to be found in the debug log, which I was able to reproduce (see debug log below):
`11:57:32.336 5110 8727 W WebSocketConnection: java.io.EOFException` 

My device has an SD card, but it is formatted as "internal memory". I.e. AFAIK apps etc. are installed on it.
In older versions, the backup worked (at least I have a backup file created in March).
Ejecting the SD card does not solve the issue.
**Device:** Moto G5
**Android version:** 7.0
**Signal version:** 4.20.4

**Link to debug log:** 
https://debuglogs.org/04acecd90af904d86bde6640d18bacbea0bd555fe5f6a2a1921df26debb2132b

**Comment by PiCob** [comment]
Those of you using 4.19 should wait for 4.20 and try again (probably it will work then)

**Comment by bme13** [comment]
Same issue here, also on a device with external sd card.
This is the logfile: https://debuglogs.org/e84fbf6f7e6e8985195f2d4c11cfe1efa950d6fa28b9781124ed739f6dcb121b
Before the line:
`java.io.IOException: Renaming temporary backup file failed!`
it also shows
`System.err: rename failed: EXDEV (Cross-device link) : /storage/emulated/0/Android/data/org.thoughtcrime.securesms/cache/backup1318601257tmp -> /storage/9016-4EF8/Android/data/org.thoughtcrime.securesms/cache/signal-2018-05-24-13-34-34.backup`
Device: Samsung SM-G900F
Android version 6.0.1
Signal version: 4.20.4
note the device is rooted.

**Comment by devkid** [comment]
Why is this closed? This bug is not fixed yet in the latest 4.20.4.

**Comment by PiCob** [comment]
@devkid: It **is** fixed (at least supposed to be) and will be included in the **next** release.

**Comment by devkid** [comment]
Sorry, my bad, didn't see the commit on the "closed this" message :(

**Comment by OTSim** [comment]
Same thing happened on my Samsung S5. Workaround: make the backup with the SD card removed.

**Comment by nextgen-networks** [comment]
something weird happend to me

I've used QtADB_0.8.1 to check the contents of "/sdcard/Android/data/org.thoughtcrime.securesms/cache/" with the file browser.
Saw there many backups that have been created since 2018-05-22 and also an (supposingly) not removed temp file from an olter backup (dated 2018-05-14).

After backup of the not moved backups to PC via QtADB_0.8.1 (and not changing anything in the cache folder!) i restarted the phone and tried a backup again ...
... and guys what to say: Backup is working again! It moves the new created backup to /sdcard/Signal/Backups/ again ...

... maybe whe have some read/write permission problems while moving the backup file from chache to regular backup folder that have been fixed while using QtADB_0.8.1?

Signal version 4.20.4
OS Android 8.0.0
Device: Xiaomi Mi MIX 2
ROM: xiaomi.eu 8.5.8 weekly
